internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+	"github.com/openshift/assisted-service/models"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// BMACReconciler reconciles a Agent object
+type BMACReconciler struct {
+	client.Client
+	Log    logrus.FieldLogger
+	Scheme *runtime.Scheme
+}
+
+const (
+	AGENT_BMH_LABEL                 = "adi.openshift.io/bmh"
+	BMH_AGENT_ROLE                  = "bmac.adi.openshift.io/role"
+	BMH_AGENT_HOSTNAME              = "bmac.adi.openshift.io/hostname"
+	BMH_AGENT_MACHINE_CONFIG_POOL   = "bmac.adi.openshift.io/machine-config-pool"
+	BMH_INSTALL_ENV_LABEL           = "installenvs.adi.openshift.io"
+	BMH_INSPECT_ANNOTATION          = "inspect.metal3.io"
+	BMH_HARDWARE_DETAILS_ANNOTATION = "inspect.metal3.io/hardwaredetails"
+)
+
+// reconcileResult is an interface that encapsulates the result of a Reconcile
+// call, as returned by the action corresponding to the current state.
+type reconcileResult interface {
+	Result() (reconcile.Result, error)
+	Dirty() bool
+}
+
+// reconcileComplete is a result indicating that the current reconcile has completed,
+// and there is nothing else to do.
+type reconcileComplete struct {
+	dirty bool
+}
+
+func (r reconcileComplete) Result() (result reconcile.Result, err error) {
+	return
+}
+
+func (r reconcileComplete) Dirty() bool {
+	return r.dirty
+}
+
+// reconcileRequeue is a result indicating that the reconcile should be
+// requeued.
+type reconcileRequeue struct {
+	delay time.Duration
+}
+
+func (r reconcileRequeue) Result() (result reconcile.Result, err error) {
+	result.RequeueAfter = r.delay
+	// Set Requeue true as well as RequeueAfter in case the delay is 0.
+	result.Requeue = true
+	return
+}
+
+func (r reconcileRequeue) Dirty() bool {
+	return false
+}
+
+// reconcileError is a result indicating that an error occurred while attempting
+// to advance the current reconcile, and that reconciliation should be retried.
+type reconcileError struct {
+	err error
+}
+
+func (r reconcileError) Result() (result reconcile.Result, err error) {
+	err = r.err
+	return
+}
+
+func (r reconcileError) Dirty() bool {
+	return false
+}
+
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;watch;update;patch
+
+func (r *BMACReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		return reconcileComplete{}.Result()
+	}
+
+	// Let's reconcile the BMH
+	result := r.reconcileBMH(bmh)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.Errorf("Error updating hardwaredetails %v", err)
```

Comment:
it is not hardwaredetails, it is relevant to url mainly, right?

Reply:
lets do log.WithError?

Reply:
+1